Let RRef getValue() synchronize CUDA streams

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56895 Let RRef getValue() synchronize CUDA streams**

PR #54932 fixes CUDA stream synchronization between RPC-created
OwnerRRef and UserRRef when `to_here()` is invoked. However, there
are two more gaps.

1. RRef value can be accessed on the owner directly through
    `local_value`, which bypasses the fix in #54932.
2. When RRef is created directly through RRef ctor instead of RPC,
    the OwnerRRef won't be able to correctly record CUDA events.

This PR fixes 1 by letting current streams wait for RRef recorded
CUDA events before returning the value in `RRef::getValue()`.

For 2, more discussions is needed to decide whether we should add
a `devices` argument to RRef ctor, or should RRef ctor inspect the
given values.

Differential Revision: [D27992775](https://our.internmc.facebook.com/intern/diff/D27992775)

mrshenli committed:
Let RRef getValue() synchronize CUDA streams

PR #54932 fixes CUDA stream synchronization between RPC-created
OwnerRRef and UserRRef when `to_here()` is invoked. However, there
are two more gaps.

1. RRef value can be accessed on the owner directly through
    `local_value`, which bypasses the fix in #54932.
2. When RRef is created directly through RRef ctor instead of RPC,
    the OwnerRRef won't be able to correctly record CUDA events.

This PR fixes 1 by letting current streams wait for RRef recorded
CUDA events before returning the value in `RRef::getValue()`.

For 2, more discussions is needed to decide whether we should add
a `devices` argument to RRef ctor, or should RRef ctor inspect the
given values.

[ghstack-poisoned]

lw commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -785,6 +787,15 @@ c10::intrusive_ptr<RRef> RRefContext::delForkOfOwner(
 
+void RRefContext::blockCurrentStreams(
+    const std::vector<c10::Event>& events) const {
+  if (currentStreamFactory_) {
+    for (const c10::Event& event : events) {
+      event.block(currentStreamFactory_(event.device_index()));
+    }
+  }
```
I think this could be also written as follows:
```suggestion
  for (const c10::Event& event : events) {
    c10::Device device{event.device_type(), event.device_index()};
    c10::Stream stream = c10::getDeviceGuardImpl(device.type())->getStream(device);
    event.block(stream);
  }
```
This way it leverages the c10 virtual device guards, which means it will work in a generic way both when CUDA is present and when it isn't. This allows us to avoid all this `currentStreamFactory_` complexity (moreover, before if `currentStreamFactory_` was null we'd silently avoid synchronizing the streams leading to correctness issues, whereas now if anything went wrong we'd see it immediately).

Here's some references on the above functions:
https://github.com/pytorch/pytorch/blob/7b31ba470897c4e5c4acb01d632ac71dd0b7d6ef/c10/core/impl/DeviceGuardImplInterface.h#L265
https://github.com/pytorch/pytorch/blob/7b31ba470897c4e5c4acb01d632ac71dd0b7d6ef/c10/core/impl/DeviceGuardImplInterface.h#L103

lw commented on torch/csrc/distributed/rpc/rref_impl.cpp:
```diff
@@ -239,6 +239,9 @@ const IValue& OwnerRRef::getValue() const {
   }
+  // Before accessing the value in this RRef, current CUDA streams must wait
+  // for pending CUDA operations that create the value.
+  RRefContext::getInstance().blockCurrentStreams(events_);
```
I know nothing about RRefs, hence let me ask: a few lines above we call `future_->wait()`, what future is that? What value does it hold? If the future is a CUDAFuture which holds the value, doesn't it already store some events on its own, and wouldn't calling `wait()` already synchronize those events?

mrshenli replied:
Currently it's an `ivalue::Future`. I do plan to convert that to a CUDAFuture on-demand when  we have a consensus on how to record devices in non-RPC RRefs

lw replied:
Let me reiterate that I know nothing about RRefs, but, if we changed the future to a CUDAFuture now could that help solve any of our problems? (current, or even past problems)

I'd find it cleaner to have all the synchronization logic in a single place (CUDAFuture) and have everything else just leverage that logic instead of reimplementing it.

mrshenli committed:
Update on "Let RRef getValue() synchronize CUDA streams"


PR #54932 fixes CUDA stream synchronization between RPC-created
OwnerRRef and UserRRef when `to_here()` is invoked. However, there
are two more gaps.

1. RRef value can be accessed on the owner directly through
    `local_value`, which bypasses the fix in #54932.
2. When RRef is created directly through RRef ctor instead of RPC,
    the OwnerRRef won't be able to correctly record CUDA events.

This PR fixes 1 by letting current streams wait for RRef recorded
CUDA events before returning the value in `RRef::getValue()`.

For 2, more discussions is needed to decide whether we should add
a `devices` argument to RRef ctor, or should RRef ctor inspect the
given values.

Differential Revision: [D27992775](https://our.internmc.facebook.com/intern/diff/D27992775)

[ghstack-poisoned]

mrshenli committed:
Update on "Let RRef getValue() synchronize CUDA streams"


PR #54932 fixes CUDA stream synchronization between RPC-created
OwnerRRef and UserRRef when `to_here()` is invoked. However, there
are two more gaps.

1. RRef value can be accessed on the owner directly through
    `local_value`, which bypasses the fix in #54932.
2. When RRef is created directly through RRef ctor instead of RPC,
    the OwnerRRef won't be able to correctly record CUDA events.

This PR fixes 1 by letting current streams wait for RRef recorded
CUDA events before returning the value in `RRef::getValue()`.

For 2, more discussions is needed to decide whether we should add
a `devices` argument to RRef ctor, or should RRef ctor inspect the
given values.

Differential Revision: [D27992775](https://our.internmc.facebook.com/intern/diff/D27992775)

[ghstack-poisoned]

lw commented on torch/csrc/distributed/rpc/rref_context.h:
```diff
@@ -206,6 +206,9 @@ class TORCH_API RRefContext {
   std::unordered_map<std::string, std::string> getDebugInfo();
 
+  // use the given Events to block current streams on corresponding devices
+  void blockCurrentStreams(const std::vector<c10::Event>& events) const;
```
Extra nit: this could be static. In fact, we don't even need this to be a method on RRefContext at all: it could just be a helper function in an anonymous namespace directly inside `rref_impl.cpp`.

mrshenli committed:
Update on "Let RRef getValue() synchronize CUDA streams"


PR #54932 fixes CUDA stream synchronization between RPC-created
OwnerRRef and UserRRef when `to_here()` is invoked. However, there
are two more gaps.

1. RRef value can be accessed on the owner directly through
    `local_value`, which bypasses the fix in #54932.
2. When RRef is created directly through RRef ctor instead of RPC,
    the OwnerRRef won't be able to correctly record CUDA events.

This PR fixes 1 by letting current streams wait for RRef recorded
CUDA events before returning the value in `RRef::getValue()`.

For 2, more discussions is needed to decide whether we should add
a `devices` argument to RRef ctor, or should RRef ctor inspect the
given values.

Differential Revision: [D27992775](https://our.internmc.facebook.com/intern/diff/D27992775)

[ghstack-poisoned]

mrshenli committed:
Update on "Let RRef getValue() synchronize CUDA streams"


PR #54932 fixes CUDA stream synchronization between RPC-created
OwnerRRef and UserRRef when `to_here()` is invoked. However, there
are two more gaps.

1. RRef value can be accessed on the owner directly through
    `local_value`, which bypasses the fix in #54932.
2. When RRef is created directly through RRef ctor instead of RPC,
    the OwnerRRef won't be able to correctly record CUDA events.

This PR fixes 1 by letting current streams wait for RRef recorded
CUDA events before returning the value in `RRef::getValue()`.

For 2, more discussions is needed to decide whether we should add
a `devices` argument to RRef ctor, or should RRef ctor inspect the
given values.

Differential Revision: [D27992775](https://our.internmc.facebook.com/intern/diff/D27992775)

[ghstack-poisoned]
